File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/canvas/CanvasServer.java
package pw.switchcraft.plethora.gameplay.modules.glasses.canvas;

import it.unimi.dsi.fastutil.ints.*;
import net.minecraft.server.network.ServerPlayerEntity;
import pw.switchcraft.plethora.api.method.IAttachable;
import pw.switchcraft.plethora.api.module.IModuleAccess;
import pw.switchcraft.plethora.api.reference.ConstantReference;
import pw.switchcraft.plethora.api.reference.IReference;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject.BaseObjectReference;
import pw.switchcraft.plethora.gameplay.modules.glasses.networking.CanvasAddPacket;
import pw.switchcraft.plethora.gameplay.modules.glasses.networking.CanvasRemovePacket;
import pw.switchcraft.plethora.gameplay.modules.glasses.networking.CanvasUpdatePacket;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler.ID_2D;
import static pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler.ID_3D;

public class CanvasServer implements ConstantReference<CanvasServer>, IAttachable {
    private final int canvasId;
    private final IModuleAccess access;
    private final ServerPlayerEntity player;

    private final Int2ObjectMap<BaseObject> objects = new Int2ObjectOpenHashMap<>();
    private final Int2ObjectMap<IntSet> childrenOf = new Int2ObjectOpenHashMap<>();

    private final IntSet removed = new IntOpenHashSet();

    private final AtomicInteger lastId = new AtomicInteger(ID_3D);

    private final ObjectGroup.Frame2d group2D = () -> ID_2D;
    private final ObjectGroup.Origin3d origin3D = () -> ID_3D;

    public CanvasServer(@Nonnull IModuleAccess access, @Nonnull ServerPlayerEntity player) {
        canvasId = CanvasHandler.nextId();
        this.access = access;
        this.player = player;

        childrenOf.put(ID_2D, new IntOpenHashSet());
        childrenOf.put(ID_3D, new IntOpenHashSet());
    }

    @Override
    public void attach() {
        access.getData().putInt("id", canvasId);
        access.markDataDirty();
        CanvasHandler.addServer(this);
    }

    @Override
    public void detach() {
        CanvasHandler.removeServer(this);
        access.getData().remove("id");
        access.markDataDirty();
    }

    public int newObjectId() {
        return lastId.incrementAndGet();
    }

    public ObjectGroup.Frame2d canvas2d() {
        return group2D;
    }

    public ObjectGroup.Origin3d canvas3d() {
        return origin3D;
    }

    @Nonnull
    synchronized CanvasAddPacket getAddPacket() {
        return new CanvasAddPacket(canvasId, objects.values().toArray(new BaseObject[0]));
    }

    @Nonnull
    CanvasRemovePacket getRemovePacket() {
        return new CanvasRemovePacket(canvasId);
    }

    @Nullable
    synchronized CanvasUpdatePacket getUpdatePacket() {
        List<BaseObject> changed = null;
        for (BaseObject object : objects.values()) {
            if (object.pollDirty()) {
                if (changed == null) changed = new ArrayList<>();
                changed.add(object);
            }
        }

        if (changed == null && removed.isEmpty()) return null;

        if (changed == null) changed = Collections.emptyList();
        CanvasUpdatePacket packet = new CanvasUpdatePacket(
            canvasId, changed, removed.toIntArray()
        );

        removed.clear();

        return packet;
    }

    public synchronized void add(@Nonnull BaseObject object) {
        IntSet parent = childrenOf.get(object.parent());
        if (parent == null) throw new IllegalArgumentException("No such parent");

        if (objects.put(object.id(), object) != null) {
            throw new IllegalStateException("An object already exists with that key");
        }

        parent.add(object.id());
        if (object instanceof ObjectGroup) childrenOf.put(object.id(), new IntOpenHashSet());
    }

    public synchronized void remove(BaseObject object) {
        if (!removeImpl(object.id())) {
            throw new IllegalStateException("No such object with this key");
        }
    }

    public synchronized BaseObject getObject(int id) {
        return objects.get(id);
    }

    public synchronized void clear(ObjectGroup object) {
        IntSet children = childrenOf.get(object.id());
        if (children == null) throw new IllegalStateException("Object has no children");

        clearImpl(children);
    }

    private boolean removeImpl(int id) {
        if (objects.remove(id) == null) return false;

        IntSet children = childrenOf.remove(id);
        if (children != null) clearImpl(children);

        removed.add(id);
        return true;
    }

    private void clearImpl(IntSet objects) {
        for (IntIterator iterator = objects.iterator(); iterator.hasNext(); ) {
            int childId = iterator.nextInt();
            removeImpl(childId);
            iterator.remove();
        }
    }

    @Nonnull
    public ServerPlayerEntity getPlayer() {
        return player;
    }

    @Nonnull
    @Override
    public CanvasServer get() {
        return this;
    }

    @Nonnull
    @Override
    public CanvasServer safeGet() {
        return this;
    }

    /**
     * Get a reference to this object
     *
     * @param baseObject@return The resulting reference.
     */
    public <T extends BaseObject> IReference<T> reference(T baseObject) {
        return new BaseObjectReference<>(this, baseObject);
    }
}

File: src/main/java/pw/switchcraft/plethora/gameplay/modules/glasses/canvas/CanvasServer.kt
package pw.switchcraft.plethora.gameplay.modules.glasses.canvas

import it.unimi.dsi.fastutil.ints.Int2ObjectMap
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap
import it.unimi.dsi.fastutil.ints.IntOpenHashSet
import it.unimi.dsi.fastutil.ints.IntSet
import net.minecraft.server.network.ServerPlayerEntity
import pw.switchcraft.plethora.api.method.IAttachable
import pw.switchcraft.plethora.api.module.IModuleAccess
import pw.switchcraft.plethora.api.reference.ConstantReference
import pw.switchcraft.plethora.api.reference.IReference
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler.ID_2D
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.CanvasHandler.ID_3D
import pw.switchcraft.plethora.gameplay.modules.glasses.networking.CanvasAddPacket
import pw.switchcraft.plethora.gameplay.modules.glasses.networking.CanvasRemovePacket
import pw.switchcraft.plethora.gameplay.modules.glasses.networking.CanvasUpdatePacket
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.BaseObject.BaseObjectReference
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup.Frame2d
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.ObjectGroup.Origin3d
import java.util.concurrent.atomic.AtomicInteger

class CanvasServer(
  access: IModuleAccess,
  player: ServerPlayerEntity
) : ConstantReference<CanvasServer>, IAttachable {
  private val canvasId: Int = CanvasHandler.nextId()
  private val access: IModuleAccess
  val player: ServerPlayerEntity

  private val objects: Int2ObjectMap<BaseObject?> = Int2ObjectOpenHashMap()
  private val childrenOf: Int2ObjectMap<IntSet> = Int2ObjectOpenHashMap()

  private val removed: IntSet = IntOpenHashSet()

  private val lastId = AtomicInteger(ID_3D)

  val canvas2d = object: Frame2d { override val id = ID_2D }
  val canvas3d = object: Origin3d { override val id = ID_3D }

  init {
    this.access = access
    this.player = player
    childrenOf.put(ID_2D, IntOpenHashSet())
    childrenOf.put(ID_3D, IntOpenHashSet())
  }

  override fun attach() {
    access.data.putInt("id", canvasId)
    access.markDataDirty()
    CanvasHandler.addServer(this)
  }

  override fun detach() {
    CanvasHandler.removeServer(this)
    access.data.remove("id")
    access.markDataDirty()
  }

  fun newObjectId() = lastId.incrementAndGet()

  @Synchronized
  fun makeAddPacket() = CanvasAddPacket(canvasId, objects.values.toTypedArray())

  @Synchronized
  fun makeRemovePacket() = CanvasRemovePacket(canvasId)

  @Synchronized
  fun makeUpdatePacket(): CanvasUpdatePacket? {
    var changed: MutableList<BaseObject>? = null
    for (obj in objects.values) {
      if (obj!!.pollDirty()) {
        if (changed == null) changed = mutableListOf()
        changed.add(obj)
      }
    }

    if (changed == null && removed.isEmpty()) return null
    if (changed == null) changed = mutableListOf()

    val packet = CanvasUpdatePacket(canvasId, changed, removed.toIntArray())
    removed.clear()
    return packet
  }

  @Synchronized
  fun add(obj: BaseObject) {
    val parent = childrenOf[obj.parent] ?: throw IllegalArgumentException("No such parent")
    check(objects.put(obj.id, obj) == null) { "An object already exists with that key" }
    parent.add(obj.id)
    if (obj is ObjectGroup) childrenOf.put(obj.id, IntOpenHashSet())
  }

  @Synchronized
  fun remove(obj: BaseObject) {
    check(removeImpl(obj.id)) { "No such object with this key" }
  }

  @Suppress("UNCHECKED_CAST")
  @Synchronized
  fun <T : BaseObject?> getObject(id: Int): T? = objects[id] as T?

  @Synchronized
  fun clear(obj: ObjectGroup) {
    val children = childrenOf[obj.id] ?: throw IllegalStateException("Object has no children")
    clearImpl(children)
  }

  private fun removeImpl(id: Int): Boolean {
    if (objects.remove(id) == null) return false

    val children = childrenOf.remove(id)
    children?.let { clearImpl(it) }

    removed.add(id)
    return true
  }

  private fun clearImpl(objects: IntSet) {
    val iterator = objects.iterator()
    while (iterator.hasNext()) {
      val childId = iterator.nextInt()
      removeImpl(childId)
      iterator.remove()
    }
  }

  override fun get(): CanvasServer = this
  override fun safeGet(): CanvasServer = this

  /**
   * Get a reference to this object
   *
   * @param baseObject@return The resulting reference.
   */
  fun <T : BaseObject> reference(baseObject: T): IReference<T> = BaseObjectReference(this, baseObject)
}
